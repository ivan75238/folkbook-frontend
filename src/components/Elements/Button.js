import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

//region Styled
const ButtonWrapper = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0 16px;
    border-radius: 5px;
    border: ${props => props.border};
    background: ${props => props.disabled ? "gray" : props.background || "#3366FF"};
    margin: ${props => props.margin};
    padding: ${props => props.padding};
    width: ${props => props.width || "auto"};
    height: ${props => props.height || "auto"};
    cursor: pointer;
    
    &:hover {
        box-shadow: 0 0 3px rgba(0,0,0,0.4);
    }
`;

const Label = styled.p`
    font-family: Roboto,sans-serif;
    font-size: ${props => props.fontSize || "14px"};
    color: ${props => props.textColor || "#fff"};
`;
//endregion

const Button = (props) => {
    const {
        title, background,
        textColor, fontSize,
        onClick, disabled,
        width, height,
        margin, padding, type
    } = props;

    switch (type) {
        case "fill-white":
            return (
                <ButtonWrapper width={width}
                               height={height}
                               background={"transparent"}
                               disabled={disabled}
                               border={"2px solid white"}
                               onClick={disabled ? null : onClick ? onClick : null}
                               padding={padding}
                               margin={margin}>
                    <Label fontSize={fontSize}
                           textColor={"white"}>
                        {title}
                    </Label>
                </ButtonWrapper>
            );
        default:
            return (
                <ButtonWrapper width={width}
                               height={height}
                               background={background}
                               disabled={disabled}
                               onClick={disabled ? null : onClick ? onClick : null}
                               padding={padding}
                               margin={margin}>
                    <Label fontSize={fontSize}
                           textColor={textColor}>
                        {title}
                    </Label>
                </ButtonWrapper>
            )
    }
};

Button.propTypes = {
    title: PropTypes.any,
    background: PropTypes.string,
    textColor: PropTypes.string,
    fontSize: PropTypes.string,
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
    width: PropTypes.string,
    height: PropTypes.string,
    margin: PropTypes.string,
    padding: PropTypes.string,
    type: PropTypes.string,
};

export default Button;
