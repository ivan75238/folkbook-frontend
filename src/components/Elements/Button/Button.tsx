import React from "react";
import styled from "styled-components";

//region Types
type defaultProps = {
    disabled?: boolean,
    background?: string,
    margin?: string,
    padding?: string,
    width?: string,
    height?: string
}

type buttonWrapperProps = defaultProps & {border?: string}

type labelProps = {
    textColor?: string,
    fontSize?: string
}

type PropsComponent = defaultProps & labelProps & {
    title?: string,
    onClick?: () => void,
    type?: string,
    className?: string
}
//endregion

//region Styled
const ButtonWrapper = styled.div<buttonWrapperProps>`
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0 16px;
    border-radius: 5px;
    border: ${props => props.border};
    background: ${props => props.disabled ? "gray" : props.background || "#3366FF"};
    margin: ${props => props.margin};
    padding: ${props => props.padding};
    width: ${props => props.width || "auto"};
    height: ${props => props.height || "auto"};
    cursor: pointer;
    
    &:hover {
        box-shadow: 0 0 3px rgba(0,0,0,0.4);
    }
`;

const Label = styled.p<labelProps>`
    font-family: Roboto,sans-serif;
    font-size: ${props => props.fontSize || "14px"};
    color: ${props => props.textColor || "#fff"};
`;
//endregion

const Button = (props: PropsComponent) => {
    const {title, background,
        textColor, fontSize,
        onClick, disabled,
        width, height, className,
        margin, padding, type} = props;

    const onClickEvent = () => {
        if (!disabled && onClick){
            onClick();
        }
    };

    if (type === "fill-white") {
        return (
            <ButtonWrapper width={width}
                           className={"btn fill-white " + className}
                           height={height}
                           background={"transparent"}
                           disabled={disabled}
                           border={"2px solid white"}
                           onClick={onClickEvent}
                           padding={padding}
                           margin={margin}>
                <Label fontSize={fontSize}
                       textColor={"white"}>
                    {title}
                </Label>
            </ButtonWrapper>
        );
    } else {
        return (
            <ButtonWrapper width={width}
                           className={"btn " + className}
                           height={height}
                           background={background}
                           disabled={disabled}
                           onClick={onClickEvent}
                           padding={padding}
                           margin={margin}>
                <Label fontSize={fontSize}
                       textColor={textColor}>
                    {title}
                </Label>
            </ButtonWrapper>
        )
    }
};

export default Button;
