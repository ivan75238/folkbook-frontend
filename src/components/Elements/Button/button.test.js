import React from 'react';
import Button from "./Button";
import 'jest-styled-components';
import 'jsdom-global/register';

const props = {
    title: "Test title",
    background: "black",
    textColor: "white",
    fontSize: "14px",
    onClick: () => {},
    disabled: false,
    width: "100px",
    height: "40px",
    margin: "10px",
    padding: "10px",
    type: undefined
};

const setUp = (props) => shallow(<Button {...props}/>);

describe("Тесты компонента Button", () => {

    describe("Тесты отрисовки", () => {
        it("Отрисовка компонента Button без type", () => {
            const component = shallow(<Button/>);
            expect(component).toMatchSnapshot();
        });

        it("Отрисовка компонента Button type = fill-white", () => {
            const component = shallow(<Button type={"fill-white"}/>);
            const wrapper = component.find(".fill-white");
            expect(wrapper.length).toBe(1);
        });
    });

    describe("Тесты с наличием пропсов", () => {
        const component = setUp(props);

        it("Снапшот с пропсами", () => {
            expect(component).toMatchSnapshot();
        });
    });

    describe("Тесты без пропсов", () => {
        const component = setUp({});

        it("Снапшот без пропсов", () => {
            expect(component).toMatchSnapshot();
        });
    });

    describe("Тесты обработчика нажатия", () => {

        it("Клик при disabled=true", () => {
            const onClickCallback = jest.fn();
            const component = setUp({disabled: true, onClick: onClickCallback});
            component.find(".btn").simulate("click");
            expect(onClickCallback.mock.calls.length).toBe(0)
        });

        it("Клик при disabled=false", () => {
            const onClickCallback = jest.fn();
            const component = setUp({disabled: false, onClick: onClickCallback});
            component.find(".btn").simulate("click");
            expect(onClickCallback.mock.calls.length).toBe(1)
        });
    });

    describe("Тесты Styled Component", () => {

        it('Тест border при type = fill-white', () => {
            const component = mount(<Button type={"fill-white"}/>);
            expect(component.find('Button')).toHaveStyleRule('border', "2px solid white");
        });

        it('Тест background при disabled = true', () => {
            const component = mount(<Button type={"fill-white"} disabled={true}/>);
            expect(component.find('Button')).toHaveStyleRule('background', "gray");
        });

        it('Тест background при disabled = false и background = undef и type = undef', () => {
            const component = mount(<Button disabled={false}/>);
            expect(component.find('Button')).toHaveStyleRule('background', "#3366FF");
        });
    });
});