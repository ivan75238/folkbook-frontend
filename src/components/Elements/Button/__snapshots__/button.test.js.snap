// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты компонента Button Тесты без пропсов Снапшот без пропсов 1`] = `
<Button__ButtonWrapper
  className="btn"
  onClick={[Function]}
>
  <Button__Label />
</Button__ButtonWrapper>
`;

exports[`Тесты компонента Button Тесты отрисовки Отрисовка компонента Button без type 1`] = `
<Button__ButtonWrapper
  className="btn"
  onClick={[Function]}
>
  <Button__Label />
</Button__ButtonWrapper>
`;

exports[`Тесты компонента Button Тесты с наличием пропсов Снапшот с пропсами 1`] = `
<Button__ButtonWrapper
  background="black"
  className="btn"
  disabled={false}
  height="40px"
  margin="10px"
  onClick={[Function]}
  padding="10px"
  width="100px"
>
  <Button__Label
    fontSize="14px"
    textColor="white"
  >
    Test title
  </Button__Label>
</Button__ButtonWrapper>
`;
